e2e/simple_ondatra_test.go

```diff
@@ -0,0 +1,217 @@
+	otgCfg, err := configureATEInterfaces(t, ate, ateSrc, ateDst)
+	if err != nil {
+		t.Fatalf("cannot configure ATE interfaces via OTG, %v", err)
+	}
+
+	return otgCfg
+}
+
+func mirrorAddr(t *testing.T) string {
+	mirror := ondatra.DUT(t, "mirror")
+	data := mirror.CustomData(solver.KNEServiceMapKey).(map[string]*tpb.Service)
+	m := data["mirror-controller"]
+	if m == nil {
+		t.Fatalf("cannot find mirror data, got: %v", data)
+	}
+	return net.JoinHostPort(m.GetOutsideIp(), strconv.Itoa(int(m.GetOutside())))
+}
+
+func mirrorClient(t *testing.T, addr string) (mpb.MirrorClient, func() error) {
+	conn, err := grpc.Dial(addr,
+		grpc.WithTransportCredentials(insecure.NewCredentials()),
+		grpc.WithBlock(),
+	)
+	if err != nil {
+		t.Fatalf("cannot dial mirror, got err: %v", err)
+	}
+
+	return mpb.NewMirrorClient(conn), conn.Close
+}
+
+func startMirror(t *testing.T, client mpb.MirrorClient) {
+	t.Helper()
+	mirror := ondatra.DUT(t, "mirror")
+
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+
+	startReq := &mpb.StartRequest{
+		From: mirror.Port(t, "port1").Name(),
+		To:   mirror.Port(t, "port2").Name(),
+	}
+	if _, err := client.Start(ctx, startReq); err != nil {
+		t.Fatalf("cannot start mirror client, got err: %v", err)
+	}
+}
+
+func stopMirror(t *testing.T, client mpb.MirrorClient) {
+	t.Helper()
+	mirror := ondatra.DUT(t, "mirror")
+
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+
+	stopReq := &mpb.StopRequest{
+		From: mirror.Port(t, "port1").Name(),
+		To:   mirror.Port(t, "port2").Name(),
+	}
+	if _, err := client.Stop(ctx, stopReq); err != nil {
+		t.Fatalf("cannot stop mirror client, got err: %v", err)
+	}
+}
+
+func TestMirror(t *testing.T) {
+	addr := mirrorAddr(t)
+	client, stop := mirrorClient(t, addr)
+	defer stop()
+	startMirror(t, client)
+	time.Sleep(1 * time.Second)
+	stopMirror(t, client)
+}
+
+// TestMPLS is a simple test that creates an MPLS flow between two ATE ports and
+// checks that there is no packet loss. It validates magna's end-to-end MPLS
+// flow accounting.
+func TestMPLS(t *testing.T) {
+	// Start a mirroring session to copy packets.
+	maddr := mirrorAddr(t)
+	client, stop := mirrorClient(t, maddr)
+	defer stop()
+	startMirror(t, client)
+	time.Sleep(1 * time.Second)
+	defer func() { stopMirror(t, client) }()
+
+	otgCfg := pushBaseConfigs(t, ondatra.ATE(t, "ate"))
+
+	otg := ondatra.ATE(t, "ate").OTG()
+	otgCfg.Flows().Clear().Items()
+	mplsFlow := otgCfg.Flows().Add().SetName("MPLS_FLOW")
+	mplsFlow.Metrics().SetEnable(true)
+	mplsFlow.TxRx().Port().SetTxName(ateSrc.Name).SetRxName(ateDst.Name)
+
+	mplsFlow.Rate().SetChoice("pps").SetPps(1)
+
+	// Set up ethernet layer.
+	eth := mplsFlow.Packet().Add().Ethernet()
+	eth.Src().SetChoice("value").SetValue(ateSrc.MAC)
+	eth.Dst().SetChoice("value").SetValue(ateDst.MAC)
+
+	// Set up MPLS layer with destination label.
+	mpls := mplsFlow.Packet().Add().Mpls()
```

Comment:
meta-q: is the ordering on the packet defined here by the order the MPLS labels are created (it seems to be so). And if so, is it idiomatic to have packet construction be starting from outside label going to inside label?

Reply:
OTG (the API that we're using here) says the packet headers used are defined in the order that they are specified in the test - from outer-most to inner-most. I am not sure that this is the canonical/idiomatic ordering everywhere, but it's certainly used elsewhere (e.g., `gopacket`'s list of `[]gopacket.SerializableLayer{}` is ordered from outer-most to inner-most too.

I added a comment to reflect this :-)